File: src/policy/policy-engine.ts
import { PolicyConfig, PolicyDecision, CallContext, PolicyAction, PolicyMode } from './policy-types.js';
import { Logger } from '../utils/logger.js';
import { getNormalizer } from '../utils/payload-normalizer.js';
import { isFpWhitelisted } from '../ai/fp-whitelist.js';
import { ShellTokenizer } from './shell-tokenizer.js';
import { LRUCache } from 'lru-cache';
import { resolvePolicyPrecedence } from './policy-precedence.js';
import {
  getCachedPolicyDecision,
  isPolicyEvalCacheEnabled,
  policyEvalCacheKey,
  setCachedPolicyDecision,
  shouldCachePolicyDecision,
} from './policy-eval-cache.js';
import { walkStringLeaves } from './arg-leaf-walker.js';
import { detectPromptInjection } from '../scanners/prompt-injection-detector.js';
import {
  SYNC_POLICY_STRATEGIES,
  evaluateIdempotency,
  evaluateRedisRateLimit,
  opaStrategy,
  runShadowPolicy,
  yamlRulesStrategy,
  type PolicyEngineDeps,
  type SyncEvaluateContext,
} from './strategies/index.js';

/**
 * Policy Engine — evaluates every intercepted tools/call against configured rules.
 * Supports three modes: audit (passive), warn (flag only), block (active enforcement).
 *
 * v1.2: Integrated payload normalization and semantic shell analysis layers
 * v2.1: Replaced Map with LRUCache to prevent memory leaks under sustained load
 * v2.9: Strategy-pattern pipeline under src/policy/strategies/
 */
export class PolicyEngine {
  private rules: PolicyConfig['policy']['rules'];
  private mode: PolicyMode;
  private config: PolicyConfig;
  private callCounters: LRUCache<string, { count: number; resetAt: number }> = new LRUCache({
    max: 50000,
    ttl: 60000,
    updateAgeOnGet: false,
  });
  private normalizer: ReturnType<typeof getNormalizer>;
  private shellTokenizer = new ShellTokenizer();

  private compiledPatterns: Map<string, { compiled: RegExp[]; rule: PolicyConfig['policy']['rules'][number] }[]> = new Map();
  private compiledArgPatterns: Map<string, { field: string; compiled: RegExp[]; rule: PolicyConfig['policy']['rules'][number] }[]> = new Map();

  constructor(config: PolicyConfig) {
    this.rules = config.policy.rules;
    this.mode = config.policy.mode;
    this.config = config;
    this.normalizer = getNormalizer(config.policy.unicode_strict !== false);
    this.compilePatterns();
  }

  private compilePatterns(): void {
    for (const rule of this.rules) {
      if (rule.patterns?.length) {
        try {
          const compiled = rule.patterns.map(p => new RegExp(p, 'i'));
          this.compiledPatterns.set(rule.name, [
            ...(this.compiledPatterns.get(rule.name) || []),
            { compiled, rule },
          ]);
        } catch {
          Logger.warn(`Policy: invalid regex in rule '${rule.name}' patterns — skipping pre-compilation`);
        }
      }
      if (rule.argPatterns?.length) {
        for (const ap of rule.argPatterns) {
          try {
            const compiled = ap.patterns.map(p => new RegExp(p, 'i'));
            this.compiledArgPatterns.set(rule.name, [
              ...(this.compiledArgPatterns.get(rule.name) || []),
              { field: ap.field, compiled, rule },
            ]);
          } catch {
            Logger.warn(`Policy: invalid regex in rule '${rule.name}' argPatterns — skipping pre-compilation`);
          }
        }
      }
    }
  }

  private extractLeafValues(obj: unknown): string[] {
    return walkStringLeaves(obj).map((l) => l.value);
  }

  /**
   * Anti-evasion token budget: use reported count and UTF-8 byte inflation from arguments.
   */
  private effectiveRequestTokens(ctx: CallContext): number {
    let inflated = 0;
    if (ctx.arguments) {
      for (const { value } of walkStringLeaves(ctx.arguments)) {
        inflated += Buffer.byteLength(value, 'utf8');
        for (const ch of value) {
          const cp = ch.codePointAt(0)!;
          if (cp > 0x7f) inflated += 2;
        }
      }
    }
    const byteEstimate = Math.ceil(inflated / 4);
    return Math.max(ctx.requestTokens, byteEstimate);
  }

  private buildDeps(): PolicyEngineDeps {
    return {
      config: this.config,
      rules: this.rules,
      mode: this.mode,
      normalizer: this.normalizer,
      shellTokenizer: this.shellTokenizer,
      compiledPatterns: this.compiledPatterns,
      compiledArgPatterns: this.compiledArgPatterns,
      callCounters: this.callCounters,
      resolveAction: (a) => this.resolveAction(a),
      extractLeafValues: (o) => this.extractLeafValues(o),
      evaluateRule: (rule, ctx, analysis, skip) =>
        this.evaluateRule(rule, ctx, analysis, skip),
    };
  }

  isOpaEnabled(): boolean {
    if (!process.env['OPA_URL']) return false;
    if (this.config.policy.opa === false) return false;
    return this.config.policy.opa === true || process.env['GUARDIAN_OPA_ENABLED'] === 'true';
  }

  async evaluateAsync(context: CallContext): Promise<PolicyDecision> {
    runShadowPolicy(context);

    const idempotencyDecision = await evaluateIdempotency(context, this.mode);
    if (idempotencyDecision) return idempotencyDecision;

    const deps = this.buildDeps();
    const opaDecision = this.isOpaEnabled()
      ? await opaStrategy.evaluateAsync(context, deps)
      : null;

    const { decision: rateDecision, skipLocalRateLimit } = await evaluateRedisRateLimit(context, deps);
    if (rateDecision) {
      return resolvePolicyPrecedence(opaDecision, rateDecision);
    }

    if (isPolicyEvalCacheEnabled()) {
      const cacheKey = policyEvalCacheKey(context);
      const cached = await getCachedPolicyDecision(cacheKey);
      if (cached) return resolvePolicyPrecedence(opaDecision, cached);
    }

    const yamlDecision = this.evaluate(context, { skipLocalRateLimit });
    const finalDecision = resolvePolicyPrecedence(opaDecision, yamlDecision);
    if (isPolicyEvalCacheEnabled() && shouldCachePolicyDecision(finalDecision)) {
      await setCachedPolicyDecision(policyEvalCacheKey(context), finalDecision);
    }
    return finalDecision;
  }

  /** Clear in-memory per-minute call counters (harness / isolated rate-limit suites). */
  resetRateCounters(): void {
    this.callCounters.clear();
  }

  evaluate(
    context: CallContext,
    options?: { skipLocalRateLimit?: boolean; yamlOnly?: boolean },
  ): PolicyDecision {
    const normalizedArgs = context.arguments
      ? this.normalizer.normalizeJsonValue(context.arguments) as Record<string, unknown>
      : {};
    const normalizedContext: CallContext = {
      ...context,
      arguments: normalizedArgs,
    };
    const argsStr = JSON.stringify(normalizedArgs);

    const syncCtx: SyncEvaluateContext = {
      raw: context,
      normalized: normalizedContext,
      argsStr,
      skipLocalRateLimit: options?.skipLocalRateLimit,
    };

    const deps = this.buildDeps();
    const strategies = options?.yamlOnly
      ? SYNC_POLICY_STRATEGIES.filter((s) => s === yamlRulesStrategy)
      : SYNC_POLICY_STRATEGIES;
    for (const strategy of strategies) {
      const decision = strategy.evaluate(syncCtx, deps);
      if (decision) return decision;
    }

    const defaultAction = this.config.policy.default_action ?? 'pass';
    return {
      action: this.resolveAction(defaultAction),
      rule: 'default',
      reason: `No matching rule — applying default_action: ${defaultAction}`,
    };
  }

  private evaluateRule(
    rule: PolicyConfig['policy']['rules'][number],
    ctx: CallContext,
    analysis: { argsStr: string },
    skipLocalRateLimit = false,
  ): PolicyDecision | null {
    if (rule.tools) {
      if (rule.tools.allow && rule.tools.allow.length > 0) {
        if (!rule.tools.allow.includes(ctx.toolName)) {
          if (rule.tools.enforceAllowlist) {
            return {
              action: this.resolveAction(rule.action),
              rule: rule.name,
              reason: `Tool '${ctx.toolName}' not in allowlist: [${rule.tools.allow.join(', ')}]`,
            };
          }
          return null;
        }
      }
      if (rule.tools.deny && rule.tools.deny.length > 0) {
        if (rule.tools.deny.includes(ctx.toolName)) {
          return { action: this.resolveAction(rule.action), rule: rule.name, reason: `Tool '${ctx.toolName}' is explicitly denied` };
        }
      }
    }

    if (rule.toolCategories?.deny) {
      const toolLower = ctx.toolName.toLowerCase();
      const matchesCategory = rule.toolCategories.deny.some(
        (cat) => toolLower.includes(cat.toLowerCase()),
      );
      const isException = (rule.toolAllowExceptions ?? []).includes(ctx.toolName);
      if (matchesCategory && !isException) {
        return {
          action: this.resolveAction(rule.action),
          rule: rule.name,
          reason: `Tool '${ctx.toolName}' matches destructive category in rule '${rule.name}'`,
        };
      }
    }

    if (ctx.arguments) {
      const compiledAps = this.compiledArgPatterns.get(rule.name) || [];
      for (const { field, compiled, rule: r } of compiledAps) {
        if (r.name !== rule.name) continue;
        const values: string[] = field === '*'
          ? this.extractLeafValues(ctx.arguments)
          : (ctx.arguments[field] !== undefined ? this.extractLeafValues(ctx.arguments[field]) : []);
        for (const value of values) {
          for (const regex of compiled) {
            if (regex.test(value)) {
              const patternKey = `${field}:${regex.source}`;
              if (isFpWhitelisted(rule.name, patternKey)) continue;
              return {
                action: this.resolveAction(rule.action),
                rule: rule.name,
                reason: `Argument field '${field}' matches blocked pattern in rule '${rule.name}'`,
              };
            }
          }
        }
      }
    }

    if (ctx.arguments) {
      const compiledPs = this.compiledPatterns.get(rule.name) || [];
      for (const { compiled, rule: r } of compiledPs) {
        if (r.name !== rule.name) continue;
        for (const regex of compiled) {
          if (regex.test(analysis.argsStr)) {
            if (isFpWhitelisted(rule.name, regex.source)) continue;
            return { action: this.resolveAction(rule.action), rule: rule.name, reason: `Argument pattern matched in tool call (normalized)` };
          }
        }
      }
    }

    if (rule.maxTokens) {
      const effectiveTokens = this.effectiveRequestTokens(ctx);
      if (effectiveTokens > rule.maxTokens) {
        return {
          action: this.resolveAction(rule.action),
          rule: rule.name,
          reason: `Token count ${effectiveTokens} exceeds max ${rule.maxTokens}`,
        };
      }
    }

    if (rule.rbac) {
      const identity = ctx.agentIdentity;
      if (!identity) {
        return { action: this.resolveAction(rule.action), rule: rule.name, reason: `RBAC rule '${rule.name}' requires agent identity but none provided` };
      }
      if (rule.rbac.scopes && rule.rbac.scopes.length > 0) {
        const agentScopes = identity.scopes || [];
        const hasScope = rule.rbac.scopes.some((required) =>
          agentScopes.some((s) => s.toLowerCase() === required.toLowerCase()),
        );
        if (!hasScope) {
          return { action: this.resolveAction(rule.action), rule: rule.name, reason: `Agent '${identity.sub}' missing required scope. Need one of: [${rule.rbac.scopes.join(', ')}], have: [${agentScopes.join(', ') || 'none'}]` };
        }
      }
      if (rule.rbac.clientIds && rule.rbac.clientIds.length > 0) {
        const clientId = identity.clientId || '';
        const matches = rule.rbac.clientIds.some(pattern => {
          try {
            return new RegExp(pattern).test(clientId);
          } catch {
            Logger.warn(`Policy: invalid clientId regex pattern in rule '${rule.name}': ${pattern}`);
            return false;
          }
        });
        if (!matches) {
          return { action: this.resolveAction(rule.action), rule: rule.name, reason: `Client ID '${clientId}' not allowed. Allowed patterns: [${rule.rbac.clientIds.join(', ')}]` };
        }
      }
      if (rule.rbac.tenants && rule.rbac.tenants.length > 0) {
        const requestTenant = ctx.tenantId || process.env['GUARDIAN_TENANT_ID'] || 'default';
        if (!rule.rbac.tenants.includes(requestTenant)) {
          return {
            action: this.resolveAction(rule.action),
            rule: rule.name,
            reason: `Tenant '${requestTenant}' not allowed for rule '${rule.name}'. Allowed: [${rule.rbac.tenants.join(', ')}]`,
          };
        }
      }
    }

    if (rule.maxCallsPerMinute && !skipLocalRateLimit) {
      const tenant = ctx.tenantId || process.env['GUARDIAN_TENANT_ID'] || 'default';
      const clientId = ctx.agentIdentity?.clientId || ctx.agentIdentity?.sub;
      const key = clientId
        ? `${tenant}:${ctx.serverName}:${ctx.toolName}:${clientId}`
        : `${tenant}:${ctx.serverName}:${ctx.toolName}`;
      const now = Date.now();
      let counter = this.callCounters.get(key);
      if (!counter || now > counter.resetAt) {
        counter = { count: 1, resetAt: now + 60000 };
      } else {
        counter.count++;
      }
      this.callCounters.set(key, counter);
      if (counter.count > rule.maxCallsPerMinute) {
        return { action: this.resolveAction(rule.action), rule: rule.name, reason: `Rate limit exceeded: ${counter.count}/${rule.maxCallsPerMinute} calls per minute` };
      }
    }

    return null;
  }

  private resolveAction(ruleAction: PolicyAction): PolicyAction {
    if (this.mode === 'audit') return 'pass';
    if (this.mode === 'warn' && ruleAction === 'block') return 'flag';
    return ruleAction;
  }

  getMode(): PolicyMode {
    return this.mode;
  }

  getRules(): ReadonlyArray<PolicyConfig['policy']['rules'][number]> {
    return this.rules;
  }

  getRuleCount(): number {
    return this.rules.length;
  }

  private static RESPONSE_EXFILTRATION_PATTERNS: RegExp[] = [
    /\b(?:curl|wget|fetch|XMLHttpRequest|axios)\b.*\b(?:https?:\/\/[^\s"']+)/i,
    /\b(?:curl|wget)\b\s+.*(?:\b[a-zA-Z0-9][-a-zA-Z0-9]*\.(?:com|net|org|io|dev|xyz|ru|cn|tk|ml|ga|cf|gq|pw|top|club|online|site|website|space|fun|host|press|digital|world|life|co|me|us|eu|info|biz|pro|name|tv|cc|ws|fm|to|am|ai))/i,
    /\$\(\s*(?:cat|head|tail|less|strings)\s+.*(?:~\/\.ssh|~\/\.aws|\.env|\.config|id_rsa|id_ed25519|authorized_keys|known_hosts|credentials|secret)/i,
    /`[^`]*(?:cat|head|tail)\s+.*(?:~\/\.ssh|id_rsa|\.env|credentials|secret)[^`]*`/,
    /\?token=[A-Za-z0-9\-_]{20,}/i,
    /\b(?:send|post|upload|transmit)\b.*\b(?:secret|key|token|password|credential)/i,
  ];

  evaluateResponse(
    toolName: string,
    serverName: string,
    responseBody: string | null | undefined,
  ): { clean: boolean; detections: string[] } {
    void serverName;
    const detections: string[] = [];

    if (responseBody == null || typeof responseBody !== 'string') {
      return { clean: true, detections };
    }

    const injectionFindings = detectPromptInjection(toolName, responseBody);
    for (const f of injectionFindings) {
      detections.push(`Prompt injection (${f.severity}): ${f.patternId} — ${f.description}`);
    }

    for (const pattern of PolicyEngine.RESPONSE_EXFILTRATION_PATTERNS) {
      if (pattern.test(responseBody)) {
        detections.push(`Data exfiltration: response matches '${pattern.source}'`);
      }
    }

    const b64chunks = [...responseBody.matchAll(/[A-Za-z0-9+/]{100,}={0,2}/g)];
    for (const chunk of b64chunks) {
      try {
        const decoded = Buffer.from(chunk[0], 'base64').toString('utf-8');
        if (/\b(bash|sh|cmd|powershell|eval|exec|curl|wget)\b/.test(decoded)) {
          detections.push('Base64-encoded shell command detected in response');
          break;
        }
      } catch {
        // Not valid base64 — ignore
      }
    }

    return { clean: detections.length === 0, detections };
  }

  getShellTokenizer(): ShellTokenizer {
    return this.shellTokenizer;
  }
}

File: adversarial-harness/python/policy_engine/policy_engine.py
"""
Policy engine — faithful Python port of PolicyEngine.evaluate() sync pipeline.

Order: request-prompt-injection → semantic-guards (shell + tool-chain + semantic) → yaml-rules.
Async paths (ML, session data-flow, OPA) are not replicated; harness compares via Node for those.
"""

from __future__ import annotations

import json
import re
import time
from collections import defaultdict
from pathlib import Path
from typing import Any, Optional

import yaml

from .normalizer import PayloadNormalizer
from .prompt_injection import scan_tool_call_arguments
from .semantic_guards import evaluate_semantic_guards
from .shell_tokenizer import ShellTokenizer
from .tool_chain import evaluate_tool_chain_guard
from .types import CallContext, PolicyAction, PolicyDecision, PolicyMode

REPO_ROOT = Path(__file__).resolve().parents[3]
DEFAULT_POLICY = REPO_ROOT / "default-policy.yaml"


class PolicyEngine:
    def __init__(self, config: dict[str, Any]):
        self.config = config
        policy = config.get("policy", config)
        self.mode: PolicyMode = policy.get("mode", "block")
        self.rules: list[dict[str, Any]] = policy.get("rules", [])
        self.default_action: PolicyAction = policy.get("default_action", "pass")
        self.semantic_shell: bool = policy.get("semantic_shell", True) is not False
        self.unicode_strict: bool = policy.get("unicode_strict", True) is not False
        self.normalizer = PayloadNormalizer(self.unicode_strict)
        self.shell = ShellTokenizer()
        self._compiled_patterns: dict[str, list[re.Pattern[str]]] = {}
        self._compiled_arg: dict[str, list[tuple[str, list[re.Pattern[str]]]]] = {}
        self._call_counters: dict[str, tuple[int, float]] = {}
        self._compile_patterns()

    @classmethod
    def from_default_policy(cls) -> "PolicyEngine":
        data = yaml.safe_load(DEFAULT_POLICY.read_text(encoding="utf-8"))
        return cls(data)

    @staticmethod
    def _regex_from_policy_pattern(p: str) -> str:
        """PyYAML often yields extra backslash escaping vs Node js-yaml for the same YAML."""
        if "\\\\" in p:
            p = p.replace("\\\\", "\\")
        return p

    def _compile_patterns(self) -> None:
        for rule in self.rules:
            name = rule.get("name", "")
            for p in rule.get("patterns") or []:
                try:
                    self._compiled_patterns.setdefault(name, []).append(
                        re.compile(self._regex_from_policy_pattern(p), re.I),
                    )
                except re.error:
                    pass
            for ap in rule.get("argPatterns") or []:
                field = ap.get("field", "*")
                compiled = []
                for p in ap.get("patterns") or []:
                    try:
                        compiled.append(re.compile(self._regex_from_policy_pattern(p), re.I))
                    except re.error:
                        pass
                if compiled:
                    self._compiled_arg.setdefault(name, []).append((field, compiled))

    def _resolve_action(self, action: PolicyAction) -> PolicyAction:
        if self.mode == "audit":
            return "pass"
        if self.mode == "warn" and action == "block":
            return "flag"
        return action

    def _walk_leaves(self, obj: Any) -> list[str]:
        from .arg_walker import walk_string_leaves

        return [leaf.value for leaf in walk_string_leaves(obj)]

    def _evaluate_rule(
        self,
        rule: dict[str, Any],
        ctx: CallContext,
        args_str: str,
        skip_rate: bool = False,
    ) -> Optional[PolicyDecision]:
        name = rule.get("name", "")
        action = self._resolve_action(rule.get("action", "block"))

        tools = rule.get("tools") or {}
        if tools.get("allow"):
            if ctx.tool_name in tools["allow"]:
                return None
            return PolicyDecision(
                action,
                name,
                f"Tool '{ctx.tool_name}' not in allowlist",
            )
        if tools.get("deny") and ctx.tool_name in tools["deny"]:
            return PolicyDecision(action, name, f"Tool '{ctx.tool_name}' is explicitly denied")

        cats = (rule.get("toolCategories") or {}).get("deny") or []
        exceptions = rule.get("toolAllowExceptions") or []
        tool_lower = ctx.tool_name.lower()
        if any(cat.lower() in tool_lower for cat in cats) and ctx.tool_name not in exceptions:
            return PolicyDecision(
                action,
                name,
                f"Tool '{ctx.tool_name}' matches destructive category",
            )

        for field, patterns in self._compiled_arg.get(name, []):
            values = (
                self._walk_leaves(ctx.arguments)
                if field == "*"
                else self._walk_leaves(ctx.arguments.get(field))
                if ctx.arguments.get(field) is not None
                else []
            )
            for val in values:
                for pat in patterns:
                    if pat.search(val):
                        return PolicyDecision(
                            action,
                            name,
                            f"Argument field '{field}' matches blocked pattern",
                        )

        for pat in self._compiled_patterns.get(name, []):
            if pat.search(args_str):
                return PolicyDecision(action, name, "Argument pattern matched (normalized)")

        max_tokens = rule.get("maxTokens")
        if max_tokens and ctx.request_tokens > max_tokens:
            return PolicyDecision(
                action,
                name,
                f"Token count {ctx.request_tokens} exceeds max {max_tokens}",
            )

        max_cpm = rule.get("maxCallsPerMinute")
        if max_cpm and not skip_rate:
            key = f"{ctx.server_name}:{ctx.tool_name}"
            now = time.time()
            count, reset = self._call_counters.get(key, (0, now + 60))
            if now > reset:
                count, reset = 1, now + 60
            else:
                count += 1
            self._call_counters[key] = (count, reset)
            if count > max_cpm:
                return PolicyDecision(
                    action,
                    name,
                    f"Rate limit exceeded: {count}/{max_cpm}",
                )

        return None

    def _evaluate_semantic_shell(self, args_str: str) -> Optional[PolicyDecision]:
        if not self.semantic_shell or not args_str:
            return None
        ps = self.shell.detect_powershell_risk(args_str)
        if ps:
            return PolicyDecision("block", "semantic-shell-guard", ps)
        b64 = self.shell.detect_base64_pipe_shell(args_str)
        if b64:
            return PolicyDecision("block", "semantic-shell-guard", b64)
        sub = self.shell.detect_sensitive_command_substitution(args_str)
        if sub:
            return PolicyDecision("block", "semantic-shell-guard", sub)
        risk = self.shell.analyze_risk(args_str)
        if risk.has_command_substitution:
            return PolicyDecision(
                "block",
                "semantic-shell-guard",
                "Semantic: shell command substitution detected in arguments",
            )
        if risk.dangerous_commands:
            return PolicyDecision(
                "block",
                "semantic-shell-guard",
                f"Semantic: dangerous shell commands: [{', '.join(risk.dangerous_commands)}]",
            )
        return None

    def evaluate(self, ctx: CallContext, skip_local_rate_limit: bool = False) -> PolicyDecision:
        norm_args = (
            self.normalizer.normalize_json_value(ctx.arguments or {})
            if ctx.arguments
            else {}
        )
        norm_ctx = CallContext(
            server_name=ctx.server_name,
            tool_name=ctx.tool_name,
            arguments=norm_args,
            request_id=ctx.request_id,
            request_tokens=ctx.request_tokens,
            timestamp=ctx.timestamp,
            session_id=ctx.session_id,
        )
        args_str = json.dumps(norm_args, ensure_ascii=False)

        findings = scan_tool_call_arguments(norm_args)
        if findings:
            rank = {"critical": 0, "high": 1, "medium": 2}
            top = min(findings, key=lambda f: rank.get(f.severity, 9))
            return PolicyDecision(
                self._resolve_action("block"),
                "request-prompt-injection",
                f"Prompt injection: {top.pattern_id} ({top.severity})",
            )

        shell_dec = self._evaluate_semantic_shell(args_str)
        if shell_dec:
            return shell_dec

        chain = evaluate_tool_chain_guard(norm_ctx)
        if chain:
            return PolicyDecision(
                self._resolve_action(chain.action),
                chain.rule,
                chain.reason,
            )

        sem = evaluate_semantic_guards(norm_ctx)
        if sem:
            return PolicyDecision(self._resolve_action(sem.action), sem.rule, sem.reason)

        permitted = False
        for rule in self.rules:
            if (rule.get("tools") or {}).get("allow") and norm_ctx.tool_name in rule["tools"]["allow"]:
                permitted = True
            dec = self._evaluate_rule(rule, norm_ctx, args_str, skip_local_rate_limit)
            if dec:
                return dec

        if permitted:
            return PolicyDecision("pass", "allowlist", f"Tool '{norm_ctx.tool_name}' allowlisted")

        return PolicyDecision(
            self._resolve_action(self.default_action),
            "default",
            f"default_action: {self.default_action}",
        )
